### PR TITLE
Fix TestAttachClosedOnContainerStop racey test

### PR DESCRIPTION
**Pull request by vdemeester**
This test is failing once in a while on the CI, because the docker attach command might be called after the container ends. Got this [failure](https://jenkins.dockerproject.org/job/Docker-PRs/16696/console) on janky.

This is more or less the same problem as #16582 but in the integration tests 🐻.

🐸

Signed-off-by: Vincent Demeester vincent@sbr.pm

### CODE REVIEW

File: integration-cli/docker_cli_attach_unix_test.go [hunk 1]
```diff
@@ -13,29 +13,29 @@ import (
...
-		}
-		attachCmd := exec.Command(dockerBinary, "attach", id)
-		attachCmd.Stdin = tty
-		attachCmd.Stdout = tty
-		attachCmd.Stderr = tty
-
-		if err := attachCmd.Run(); err != nil {
+		if err := attachCmd.Wait(); err != nil {
 			errChan <- err
 			return
```
Comment (duglin) [inline integration-cli/docker_cli_attach_unix_test.go]: I think this `return` can be removed.

Comment (vdemeester) [inline integration-cli/docker_cli_attach_unix_test.go]: true :)

**Comment by duglin** [comment]
I think this test is assuming the order in which things happen is:
- start container/sleep
- start attach
- sleep ends/container dies
- attach exits
However, I think things are still a bit racey because the sleep/container may complete before the attach ever starts, and in fact the change from "sleep 2" to "sleep 1" might make that happen more often. But, I don't think the issue is the sleep itself.

I think the safer approach is to make it so that the container waits for some kind of signal from the client/host that it should exit, and we don't make that happen until after the attach has started.  We've done this in other tests before but I can't seem to find an example right now.  What if we did this:
- start a container with 'bash' (will need to setup a tty)
- start attach
- send 'exit' to container via the tty, this should make the container die
- attach should exit

**Comment by vdemeester** [comment]
> However, I think things are still a bit racey because the sleep/container may complete before the attach ever starts, and in fact the change from "sleep 2" to "sleep 1" might make that happen more often. But, I don't think the issue is the sleep itself.

Yeah, I change the sleep to 1 to gain some time and, somehow to see make tight a bit more the time (and see if it's still racy). You're right, even by issuing it outside the goroutines, there is still no guarantee the attach will start before the end of the container (it all depends on how quick the daemon is).

> I think the safer approach is to make it so that the container waits for some kind of signal from the client/host that it should exit, and we don't make that happen until after the attach has started. We've done this in other tests before but I can't seem to find an example right now. What if we did this:
>
> - start a container with 'bash' (will need to setup a tty)
> - start attach
> - send 'exit' to container via the tty, this should make the container die
> - attach should exit

I like that :smile_cat:, and I also think I've seen it somewhere. I'll look for it :)

**Comment by vdemeester** [comment]
@duglin updated. I think it's closer to the safe approach than before :)

**Comment by duglin** [comment]
interesting use of the `--stop-signal` :-)
LGTM

**Comment by cpuguy83** [comment]
Why stop-signal and the for loop? Would `top` + `docker stop` not work?

**Comment by vdemeester** [comment]
@cpuguy83 because it was fancier :stuck_out_tongue_closed_eyes:. Yeah this would work and be easier to read ; and better because it will not depend on another docker feature :smile_cat:.

Updated to not use `--stop-signal`, but still with the loop.

**Comment by cpuguy83** [comment]
LGTM

**Comment by duglin** [comment]
LGTM just waiting for janky

**Comment by duglin** [comment]
Failure isn't related.

**Pull request merged**